internal/persistence/ent/client/client.go

```diff
@@ -1,22 +1,90 @@
 package entclient
 
 import (
+	"context"
+	"time"
+
 	"github.com/go-logr/logr"
+	"github.com/google/uuid"
 
+	goapackage "github.com/artefactual-sdps/enduro/internal/api/gen/package_"
+	"github.com/artefactual-sdps/enduro/internal/event"
+	"github.com/artefactual-sdps/enduro/internal/package_"
 	"github.com/artefactual-sdps/enduro/internal/persistence"
 	"github.com/artefactual-sdps/enduro/internal/persistence/ent/db"
 )
 
 type client struct {
 	logger logr.Logger
+	evsvc  event.EventService
 	ent    *db.Client
 }
 
 var _ persistence.Service = (*client)(nil)
 
-func New(logger logr.Logger, ent *db.Client) persistence.Service {
+func New(logger logr.Logger, evsvc event.EventService, ent *db.Client) persistence.Service {
```

Comment:
This could be a good use case for the decorator pattern to decouple the concerns of persisting the data and publishing the events. I'm curious to know if you think it would be a good fit. You'd have to create an additional implementation of `persistence.Service` that wraps the `client`. In each method you would invoke the corresponding method in `client` and emit the appropiate event based on the outcome.

```go
type eventManager struct {
	evsvc event.EventService
	inner Service
}

var _ Service = (*eventManager)(nil)

func WithEvents(evsvc event.EventService, inner Service) *eventManager {
	return &eventManager{evsvc, inner}
}

// You can always use goimpl to generate large interface stubs automatically.
func (m *eventManager) CreatePackage(ctx context.Context, pkg *package_.Package) (int, error) {
	// _, err := m.inner.CreatePackage(...)
	// ... publish event if err is nil
}
```

There is an example in sdps_preprocessing, see `pkg/persistence/wrapper.go`.

Reply:
Nice! I implemented the eventManager decorator and I am quite happy with the results. :)

Reply:
That's awesome.